[MRG + 1] Add test for __dict__

kiote opened the pull request:
#### Reference Issue

see #7297
#### What does this implement/fix? Explain your changes.

Check that estimator does not change the state of **dict** during `fit` phase.
#### Any other comments?

I add "WIP = work in progress" for this PR, since I'm not sure it does what it should do. So please give me some feedback first :)

After this, I'll add checking for `transform` stage as well.

kiote committed:
Add test for __dict__ for estimator checks

  check that "predict", "transform", "decision_function" or
"predict_proba" methods do not change the state of __dict__ of any
estimator

  see #7297

amueller commented:
We certainly change the ``__dict__`` during ``fit``. The issue is not to change it during any other method.

kiote commented:
Got it. Now I'm checking that `__dict__` is not changing during `transform`.
But that made me have a lot of questions.

1. I can see that not all estimators have `transform` method. For now, I just check if it has, is it okay?
2. I see a lot of deprecation warning like this: "DeprecationWarning: Function transform is deprecated; Support to use estimators as feature selectors will be removed in version 0.19. Use SelectFromModel instead."  Which made me think I shouldn't use `transform` anymore... Right?
3. [Here](https://github.com/scikit-learn/scikit-learn/pull/7553/files#diff-b6e2c29089c86706dc868c7bfd58c1d6R259) I had to skip some estimators, because data doesn't fit, or who knows why. Seems like dirty hack, not sure how should I do that in correct way.

TomDLT commented on sklearn/tests/test_common.py:
```diff
@@ -36,6 +36,8 @@
     check_get_params_invariance,
     _set_test_name)
 
+from sklearn.datasets import load_iris, load_boston
```
load_boston is not used

amueller commented:
You should check any of ``transform``, ``predict``, ``predict_proba`` and ``decision_function``. I don't think there's a model that has all of these, but you should check any of them if they are available (starting with ``transform`` is fine). You can just ignore the deprecation warnings. Check out how that is done in ``check_estimators.py``. Why were the ones failing that you are passing? These are surprising to me.

jnothman commented:
(Before deprecation of the feature selection transformation, there certainly should have been models that have all of these, such as logistic regression)

kiote commented:
okay, what I've done by now:

1. add `check_dict` method to `estimator_checks`
2. call this method in `test_common` for (almost) all estimators.
3. still need to skip this estimators: `['SpectralCoclustering', 'SpectralBiclustering', '_BinaryGaussianProcessClassifierLaplace']`. I think it's because I don't really know how to prepare data for them, so want to fix this skip soon.
4. Have some merge conflict, want to fix that as well.

Can you please tell if it's closer to the desired result or not? Thanks!

kiote commented:
I've added some special cases for `RANSACRegressor` and `SpectralBiclustering` in the new check. but still need to skip `SpectralCoclustering` because of 

> ValueError: Found array with 0 feature(s) (shape=(23, 0)) while a minimum of 1 is required.

error. Any suggestions?

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -397,6 +398,39 @@ def check_dtype_object(name, Estimator):
+def check_dict(name, Estimator):
+    rnd = np.random.RandomState(0)
+    if name in ['SpectralCoclustering']:
+        return 0
```
just "return", no 0.

The issues with `SpectralCoclustering` and `SpectralBiclustering` may be resolved in #6141

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -397,6 +398,39 @@ def check_dtype_object(name, Estimator):
 
 
 @ignore_warnings
+def check_dict(name, Estimator):
```
Maybe `check_dict_unchanged`

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -397,6 +398,39 @@ def check_dtype_object(name, Estimator):
+    if name in ['RANSACRegressor']:
+        X = 3 * rnd.uniform(size=(20, 3))
+    else:
+        X = 2 * rnd.uniform(size=(20, 3))
```
We can't use `3 *` in all cases?

kiote replied:
If I use `3 * ` for all, I get `ValueError: _BinaryGaussianProcessClassifierLaplace supports only binary classification. y contains classes [0 1 2]`. Which is sounds fair enough.

With `2 *` for all I get on the other side `ValueError: No inliers found, possible cause is setting residual_threshold (None) too low.` for `RANSACRegressor ` :(

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -397,6 +398,39 @@ def check_dtype_object(name, Estimator):
+    set_testing_parameters(estimator)
+
+    if hasattr(estimator, "n_components"):
+        estimator.n_components = 3
```
why 3? other tests with this parameter setting use 1.

kiote replied:
well yes, but for some reason with `1` I catch `ValueError: n_best cannot be larger than n_components, but 3 >  1` from `SpectralBiclustering`

amueller replied:
Hm maybe SpectralBiclustering needs ``n_best=1``?

kiote replied:
thanks, that helped!

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -397,6 +398,39 @@ def check_dtype_object(name, Estimator):
+    estimator = Estimator()
+    set_testing_parameters(estimator)
+
+    if hasattr(estimator, "n_components"):
```
*Hmm... I wonder if these should be in `set_testing_parameters`)

kiote replied:
Can't say anything here, I've seen the same part on several places, for example here: https://github.com/kiote/scikit-learn/blob/feature-test-__dict__/sklearn/utils/estimator_checks.py#L443

So, it looks like it could be moved there, do you think it should?

amueller replied:
yeah feel free to move it there.

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -397,6 +398,39 @@ def check_dtype_object(name, Estimator):
+    for method in ["predict", "transform", "decision_function",
+                   "predict_proba"]:
+        if hasattr(estimator, method):
+            dict_before = estimator.__dict__
```
You need to perform `.copy()` at the end of this. Otherwise, in almost all cases, testing equivalence doesn't test that `__dict__` is unchanged, as you're actually comparing an object to itself.

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -397,6 +398,39 @@ def check_dtype_object(name, Estimator):
+        if hasattr(estimator, method):
+            dict_before = estimator.__dict__
+            getattr(estimator, method)(X)
+            assert_equal(estimator.__dict__, dict_before)
```
use `assert_dict_equal`

jnothman commented:
Thanks for working on this, though I suspect you will find many more failures when you correctly perform `.copy()`

amueller commented on sklearn/utils/estimator_checks.py:
```diff
@@ -398,10 +399,11 @@ def check_dtype_object(name, Estimator):
     rnd = np.random.RandomState(0)
-    if name in ['SpectralCoclustering']:
-        return 0
+    if name in ['SpectralCoclustering', 'NMF', 'ProjectedGradientNMF']:
```
How do they change the dict? Please leave them in so we can see the error and fix them.

amueller commented on sklearn/tests/test_common.py:
```diff
@@ -231,3 +232,13 @@ def test_get_params_invariance():
+
+
+@ignore_warnings
+def test_predict_does_not_change_estimator_state():
```
I think you should add this test to the generator that generates all tests in ``estimator_checks.py`` and not add a test here. Though for working on it this might be easier.

kiote replied:
got it, I remove this from here afterwards then!

kiote commented on sklearn/utils/estimator_checks.py:
```diff
@@ -397,6 +407,34 @@ def check_dtype_object(name, Estimator):
+
+    set_random_state(estimator, 1)
+
+    if name in ['SpectralBiclustering']:
```
yep! those changes helped, so I suppose I could remove this after #6141 merge, right?

kiote commented:
Not really sure, what should I do with failing tests for `NMF` and `ProjectedGradientNMF` estimators. Should I try to fix them in the scope of current issue as well?

jnothman commented:
> what should I do with failing tests for NMF and ProjectedGradientNMF estimators. Should I try to fix them in the scope of current issue as well?

It would be acceptable to skip them for now (by checking for their name) and then creating a new issue for this to be solved.

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -312,6 +314,14 @@ def set_testing_parameters(estimator):
         if not isinstance(estimator, ProjectedGradientNMF):
             estimator.set_params(solver='cd')
 
+    if hasattr(estimator, "n_components"):
```
It seems setting these parameters here is a bad idea. It severely affects other tests.

kiote replied:
Okay, I'll move that back to the concrete test. 

How did you check that, btw?

jnothman replied:
Click "details" next to "continuous-integration/travis-ci/pr" under "Some changes were not successful"

kiote replied:
removed

kiote replied:
oh I see, thanks!

jnothman reviewed (changes requested):
Getting there!

kiote commented:
I added some comments right in the code to explain some things, also I'm gong to remove new code from `tests/test_common.py` when you approve this

jnothman commented:
This looks okay... could you change the title from WIP to MRG and put the code into mergeable form?

amueller commented on sklearn/utils/estimator_checks.py:
```diff
@@ -397,6 +399,52 @@ def check_dtype_object(name, Estimator):
+def check_dict_unchanged(name, Estimator):
+    # these two estimators change the state of __dict__
+    # and need to be fixed
+    if name in ['NMF', 'ProjectedGradientNMF']:
```
Actually, it might be simpler to remove [this line](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/decomposition/nmf.py#L1109) instead, that should get rid of the issue.

kiote commented:
done!

amueller commented:
can you please rebase? Otherwise LGTM.

kiote commented:
okay! now it's rebased from master

kiote commented:
hmm, could you please tell more, what exactly do you mean? Can't see what's wrong there 😕

amueller commented:
@kiote never mind, I didn't have enough coffee...

amueller commented:
@jnothman wanna have another look?

amueller commented:
Can you please add a test to ``test_check_estimator`` that shows that ``check_estimator`` fails on an estimator that violates this?

kiote committed:
Add a test to test_check_estimator

  that shows that check_estimator fails on an estimator that violates this

kiote commented:
done, please see [here](https://github.com/scikit-learn/scikit-learn/pull/7553/commits/ba6e0939a0723eab7dcfece4a26293276d362a77)

jnothman commented on sklearn/decomposition/nmf.py:
```diff
@@ -1106,7 +1106,6 @@ def transform(self, X):
             nls_max_iter=self.nls_max_iter, sparseness=self.sparseness,
             beta=self.beta, eta=self.eta)
 
-        self.n_iter_ = n_iter_
```
Very awkwardly, there is an `Attributes` section in the docstring above (and similar ones in the docs for `fit` and `fit_transform`!). Please remove it. I suppose this change should also be documented ("Fixed bug where `NMF`'s `n_iter_` attribute was set by calls to `transform`"), though it's a very strange feature.

kiote replied:
I removed all `self.n_iter` = something in `transform`, but not really sure where this change should be documented. Could you please point me to the right place for that?

jnothman replied:
Our changelog is in `doc/whats_new.rst`

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -409,6 +411,51 @@ def check_dtype_object(name, Estimator):
+            getattr(estimator, method)(X)
+            assert_dict_equal(estimator.__dict__, dict_before,
+                              ('Estimator changes __dict__ during'
+                               'predict, transform, decision_function'
```
Could we just specify the one that's being tested?

jnothman replied:
Or would we rather list all to make it clear what the API violation is?

kiote replied:
The idea was to list them all, even if only one violates the new rule, but it might be clearer to point directly to the one that's being tested.

kiote replied:
Changed here: https://github.com/scikit-learn/scikit-learn/pull/7553/commits/4bd07c7fee4e2ba6c704421c055fed3dfa7da36e

jnothman reviewed (changes requested):
Otherwise LGTM

kiote committed:
Fixed bug where NMF's n_iter_ attribute was set by calls to transform

kiote commented on sklearn/decomposition/nmf.py:
```diff
@@ -988,6 +988,7 @@ def __init__(self, n_components=None, init=None, solver='cd',
         self.l1_ratio = l1_ratio
         self.verbose = verbose
         self.shuffle = shuffle
+        self.n_iter_ = 1
```
without having this attribute here I got 

> File "/home/travis/sklearn_build_latest/scikit-learn/sklearn/utils/estimator_checks.py", line 1600, in check_transformer_n_iter
> assert_greater_equal(estimator.n_iter_, 1)
> AttributeError: 'ProjectedGradientNMF' object has no attribute 'n_iter_'

from tests

jnothman replied:
You can remove this now.

jnothman commented on sklearn/decomposition/nmf.py:
```diff
@@ -1021,9 +1021,6 @@ def fit_transform(self, X, y=None, W=None, H=None):
         components_ : array-like, shape (n_components, n_features)
             Factorization matrix, sometimes called 'dictionary'.
 
-        n_iter_ : int
```
I was commenting that there should be no Attributes section *at all* in a method docstring. it belongs in the class docstring.

jnothman commented on sklearn/decomposition/nmf.py:
```diff
@@ -1049,7 +1046,6 @@ def fit_transform(self, X, y=None, W=None, H=None):
 
         self.n_components_ = H.shape[0]
         self.components_ = H
-        self.n_iter_ = n_iter_
```
You should not have removed this. We need it in `fit_transform`, not in `transform`

jnothman replied:
:)

jnothman reviewed (changes requested):
Sorry for the misunderstandings.

kiote commented:
okay, now I put back some lines I removed accidentally with `n_iter_`, also removed docstrings you pointed to and add documentation part about this change (mostly about NMF change).

I wonder should I add doc about the new check as well?

jnothman commented on sklearn/decomposition/nmf.py:
```diff
@@ -1066,9 +1059,6 @@ def fit(self, X, y=None, **params):
         components_ : array-like, shape (n_components, n_features)
```
Please drop all attributes here too

jnothman commented on doc/whats_new.rst:
```diff
@@ -94,6 +94,11 @@ Bug fixes
      (`#7680 <https://github.com/scikit-learn/scikit-learn/pull/7680>`_).
      By `Ibraim Ganiev`_.
 
+   - Remove params changing inside of `transform` method of :class:`decomposition.NMF`
```
The following would suffice:

```
    - Fixed a bug where :class:`decomposition.NMF` sets its `n_iters_`
      attribute in `transform()`. :issue:`7553` by `Ekaterina Krivich`_.
```

And under Enhancements, something like "check_estimator now attempts to ensure that methods `transform`, `predict`, etc. do not set attributes on the estimator."

jnothman reviewed (changes requested):
Otherwise LGTM.

jnothman commented on doc/whats_new.rst:
```diff
@@ -69,6 +69,11 @@ Enhancements
 
+   - ``check_estimator`` now attempts to ensure that methods transform, predict, etc.
+     do not set attributes on the estimator.
+     (`#7553 <https://github.com/scikit-learn/scikit-learn/pull/7553>`_)
```
again, please use
```
:issue:`7533`
```

kiote replied:
okay, anyway nor my version, not this one is not clickable for me (looks like https://github.com/scikit-learn/scikit-learn/blob/master/doc/whats_new.rst#id17). Not the scope of this issue, obviously. I'll fix that!

jnothman commented on doc/whats_new.rst:
```diff
@@ -94,10 +99,8 @@ Bug fixes
-     (`#7553 <https://github.com/scikit-learn/scikit-learn/pull/7553>`_). Since it violates
-     new represented rule "estimator state does not change at transform/predict/predict_proba time".
-     By `Ekaterina Krivich`_.
+   - Fixed a bug where :class:`decomposition.NMF` sets its `n_iters_`
```
to format `n_iters_` correctly, you need double-backticks:
```
``n_iters_``
```

jnothman commented:
Sigh. Can you update your master, rebase or merge, and fix conflicts in whats_new?

kiote commented:
oh, it's done

jnothman commented:
Thanks!
